### PR TITLE
add front-end provided context to callbacks

### PR DESCRIPTION
**Pull request by x1b6e6**
More info:
https://git.archlinux.org/pacman.git/commit/?id=e7fa35baa22bf6710a904815456d3ff679005fc7

### CODE REVIEW

**Comment by Morganamilo** [comment]
Again, you need to split this up into _five.go and _six.go files. Not that I'm exactly sure how to do that with .c files in go.

Also, there's always been a "bug" with the callbacks where they're global to all alpm instances. With context it's now possible to nicely solve that.

**Comment by x1b6e6** [comment]
> Again, you need to split this up into _five.go and _six.go files. 

Sorry for this.

> Not that I'm exactly sure how to do that with .c files in go.

I found this solution:
```go
// +build: six

package alpm

// #cgo CFLAGS: -DSIX=1
...
```

> Also, there's always been a "bug" with the callbacks where they're global to all alpm instances. With context it's now possible to nicely solve that.

I'll try implement cb work with new contexts.

**Comment by Morganamilo** [comment]
> I'll try implement cb work with new contexts.

Basically what you want to make a shim c function that takes what function as part of the context.

```go
type OurLogCtx struct  {
    Cb fn(message, level, ctx)
    our_ctx Interface
```
}

```c
logcb_shim(message, level, ctx) {
    f = ctx.Fn;
    our_ctx = ctx.our_ctx;
    f(message, level, out_ctx);
}

```go 
func (h *Handle) SetLogCallback(cb logCallbackSig, ctx Interface) {
        ctx := OurLogCtx { Cb: cb, Ctx: Interface }'
	C.alpm_set_lobcb(h.ptr, logcb_shim, &ctx);
}
```

File: callbacks_six.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#include "callbacks_six.h"
+*/
+import "C"
+
+import (
+	"unsafe"
+)
+
+type (
+	logCallbackSig      func(LogLevel, string)
```
Comment (Morganamilo) [inline callbacks_six.go:16]: This should be a struct containing the function and the context we want to pass in. So we can do:

```go
func my_log(LogLevel, string, ctx Interface) {
    //stuff
}

handle.SetLogCallback(my_log, value)
```

**Comment by x1b6e6** [comment]
I have implemented this via `func(interface{}, <other args>)`.

I'm storing context in `map[C.uint]interface{}` in Go and pass incremental id to C. This variant already working but i need help with adding tests to confirm it.

**Comment by Morganamilo** [comment]
Why do you need to store them in a map? You can just make the context a struct with the cb to call and the user context. Then in the c function call `ctx.cb(..., ctx.ctx)`.

**Comment by Jguer** [comment]
Meanwhile tests with pacman-git have been merged, should make it simpler to keep the two versions working.

I'll changing the builder image so it's a bit faster

**Comment by x1b6e6** [comment]
@Morganamilo because I can pass interfaces to C.

```
Go code may pass a Go pointer to C provided the Go memory to which it 
points does not contain any Go pointers. The C code must preserve this 
property: it must not store any Go pointers in Go memory, even temporarily. 
When passing a pointer to a field in a struct, the Go memory in question is 
the memory occupied by the field, not the entire struct. When passing a 
pointer to an element in an array or slice, the Go memory in question is the 
entire array or the entire backing array of the slice. 
```
Source:
https://stackoverflow.com/questions/51483032/passing-go-pointer-to-cgo

**Comment by Morganamilo** [comment]
It feels silly to have to store them in a map but if that's the only decent way to get it to work that's fine. Looking at the code the API is how I want it. Can't comment on if it works or is ub though.

**Comment by x1b6e6** [comment]
> It feels silly to have to store them in a map

I tried versions with struct as you described

v1:
```c
struct {
    void *go_cb;
    void *go_ctx;
};
```
v2:
```go
type struct {
    cb logCallbackSig
    ctx interface{}
}
```

Both variants is compiling but falling in runtime at call `C.go_alpm_set_log_callback` or `C.go_alpm_set_question_callback` with `panic: runtime error: cgo argument has Go pointer to Go pointer`. I think it's because interface stores another Go pointers.

**Comment by Jguer** [comment]
Seems that we're hitting some language interoperability limitations here, I think we can go with what works for now and in due time keep looking for a different solution (if this one proves to be too constraining) @x1b6e6

**Comment by Morganamilo** [comment]
What if instead of a global context pool you just had variables to store them in the handle?

**Comment by x1b6e6** [comment]
@Morganamilo if you mean go `Handle` in this form:
```go
type Handle struct {
    ptr unsafe.Pointer
    ctx_log interface{}
    ctx_question interface{}
}
```
then we can't pass it to the C context because `Go pointer to Go pointer`. Or we can pass `ptr` to C but then we need to get the context by this pointer, and this also requires the global pool of handles.

Also I can change code to using `ptr` as key in a map instead of an incremental number. This will fix freeing of the contexts after next callback changes.

**Comment by Morganamilo** [comment]
Okay, I'll leave it to you then. Plus the public API is good so if there's every a better way it can be implemented without breaking things.

**Comment by x1b6e6** [comment]
I tried build `yay` with current variant and with `ptr`. Both variants sometimes crash `yay`.

I'm looking for a solution.

[demo](https://asciinema.org/a/6ElxicRinzzr5J7KJzCe9Burd?autoplay=1)

**Comment by x1b6e6** [comment]
So crash is fixed (with `*(void**)go_cb` and `&ctx->go_cb`).

I leave variant with `h.ptr`.

I think the work is done.

**Comment by Morganamilo** [comment]
Could you rebase it on master for the CI?

**Comment by x1b6e6** [comment]
@Morganamilo done.

**Comment by Morganamilo** [comment]
API looks done and the CI passes. My go is too rust to actually review more :P

**Comment by x1b6e6** [comment]
Oops, code duplication in C.

> API looks done and the CI passes. My go is too rust to actually review more :P

I never write in go before.

**Comment by Jguer** [comment]
> API looks done and the CI passes. My go is too rust to actually review more :P

I'll take a look at it, cgo is really a bit of an edge case on the go language in terms of best practices

**Comment by Jguer** [comment]
@x1b6e6 merged, will push an updated version

**Pull request merged**